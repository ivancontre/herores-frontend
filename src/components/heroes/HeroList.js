import React, { useMemo } from 'react';
import { getHeroesByPublisher } from '../../selectors/getHerosByPublisher';
import { HeroCard } from './HeroCard';


export const HeroList = ({ publisher }) => {

    //const heroes = getHeroesByPublisher(publisher);

    const heroes = useMemo(() => getHeroesByPublisher(publisher), [ publisher ]);

    return (
        <div className="card-columns animate__animated animate__fadeIn">
            {
                heroes.map( hero => (
                    // con { ...hero } le paso todos los parametros desestructurados al componente
                    <HeroCard key={ hero.id } { ...hero } />
                ))
            }
        </div>
    )
};