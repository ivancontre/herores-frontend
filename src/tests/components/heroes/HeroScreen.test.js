import { mount } from "enzyme";
import React from 'react';
import { MemoryRouter, Route } from "react-router-dom";
import { HeroScreen } from "../../../components/heroes/HeroScreen";

// MemoryRouter: Se utiliza en las pruebas cuando queremos usar un useParams o useHistory para acceder a los datos de los parametros
// Si el componente a probar no tiene history en sus props entonces se debe falsear su history
describe('Pruebas en <HeroScreen />', () => { 
    

    test('debe de mostrarse Redirect si no hay argumentos en el URL', () => {

        const historyMock = {
            length: 10,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen history={ historyMock } />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();

        expect(wrapper.find('Redirect').exists()).toBe(true)
    })

    test('debe de mostrar un hero si el parÃ¡metro existe y se encuentra', () => {

        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroId" component={ HeroScreen } />
            </MemoryRouter>
        )

        expect(wrapper.find('.row').exists()).toBe(true)

    })

    test('debe de regresar a la pantalla anterior con PUSH', () => {

        const historyMock = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroId" component={ (props) => <HeroScreen history={ historyMock } /> } />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();

        expect(historyMock.push).toHaveBeenCalled();

        expect(historyMock.goBack).not.toHaveBeenCalled();
        
    })
    
    test('debe de regresar a la pantalla anterior con goBack', () => {

        const historyMock = {
            length: 10,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroId" component={ (props) => <HeroScreen history={ historyMock } /> } />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();

        expect(historyMock.push).not.toHaveBeenCalled();

        expect(historyMock.goBack).toHaveBeenCalled();
        
    })
    

    test('debe de llamar el redirect si el hero no existe', () => {
        
        const historyMock = {
            length: 10,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider1212']}>
                <Route path="/hero/:heroId" component={ (props) => <HeroScreen history={ historyMock } /> } />
            </MemoryRouter>
        );

        expect(wrapper.text()).toBe('')
    })
    
})
