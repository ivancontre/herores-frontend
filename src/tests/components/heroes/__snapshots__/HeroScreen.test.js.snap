// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroScreen /> debe de mostrarse Redirect si no hay argumentos en el URL 1`] = `""`;
